### PR TITLE
Add support for OSC 8 hyperlinks (HTML-like anchors)

### PR DESCRIPTION
**Pull request by geo-stark**

### CODE REVIEW

**Comment by cwendling** [comment]
See also #175

**Comment by geo-stark** [comment]
> See also #175
My patches are just backported from gnome made by the author of the comment  #175

**Comment by lukefromdc** [comment]
I have no idea how to test this, would need to be able to do so without connection to unknown/unsafe sites

**Comment by geo-stark** [comment]
> I have no idea how to test this, would need to be able to do so without connection to unknown/unsafe sites

Overall information about this feature: https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda
some simple tests:
1) printf '\e]8;;https://ubuntu.com/\e\\Ubuntu 21.10\e]8;;\e\\\n'
2) ls --hyperlink=always

**Comment by lukefromdc** [comment]
I don't really understand this, so will leave it for other team members to work on. I DO see a possible issue in your linked explanation though: autocompleting URLs could lead people to sites they don't ever want to connect to (e.g Facebook, Tiktok) or even to malware sites. Anything networked requires serious security, and coding it is out of my experience beyond blocking such sites on my machines.

**Comment by zhuyaliang** [comment]
@geo-stark Thank you for your work. I am very interested in this new function. I will review the code when I have enough time

File: src/terminal-screen.h [hunk 1]
```diff
@@ -35,7 +35,7 @@ typedef enum
     FLAVOR_VOIP_CALL,
     FLAVOR_EMAIL,
     FLAVOR_SKEY
-} TerminalURLFlavour;
+} TerminalURLFlavor;
```
Comment (vkareh) [inline src/terminal-screen.h:38]: If you're renaming this enum to make it consistent with its values, can you do it in a separate commit?

File: src/terminal-util.c [hunk 2]
```diff
@@ -707,6 +708,55 @@ terminal_util_add_proxy_env (GHashTable *env_table)
...
+	if (uri == NULL)
+		return NULL;
+
+	unesc = g_uri_unescape_string(uri, NULL);
+	if (unesc == NULL)
+		unesc = g_strdup(uri);
+
+	if (g_ascii_strncasecmp(unesc, "ftp://", 6) == 0 ||
+		g_ascii_strncasecmp(unesc, "http://", 7) == 0 ||
+		g_ascii_strncasecmp(unesc, "https://", 8) == 0)
```
Comment (vkareh) [inline src/terminal-util.c:734]: Wondering whether `mailto:` addresses can be IDN. Probably an overkill to add more.

File: src/terminal-screen.c [hunk 3]
```diff
@@ -1684,33 +1690,57 @@ terminal_screen_button_press (GtkWidget      *widget,
...
+	if (hyperlink != NULL &&
+		(event->button == 1 || event->button == 2) &&
+		(state & GDK_CONTROL_MASK))
+	{
+		gboolean handled = FALSE;
 
-	if (matched_string != NULL &&
+		g_signal_emit (screen, signals[MATCH_CLICKED], 0,
+	                 hyperlink,
+	                 FLAVOR_AS_IS,
```
Comment (vkareh) [inline src/terminal-screen.c:1710]: Since you pre-assigned `url_flavor` to `0` (which is `FLAVOR_AS_IS`), maybe send the `url_flavour` here to keep it consistent. You will override it anyway if it's not handled here.

File: src/terminal-screen.c [hunk 4]
```diff
@@ -1684,33 +1690,57 @@ terminal_screen_button_press (GtkWidget      *widget,
 	TerminalScreen *screen = TERMINAL_SCREEN (widget);
 	gboolean (* button_press_event) (GtkWidget*, GdkEventButton*) =
 	    GTK_WIDGET_CLASS (terminal_screen_parent_class)->button_press_event;
-	char *matched_string;
-	int matched_flavor = 0;
+	char *hyperlink;
+	char *url;
+	int url_flavor = 0;
```
Comment (vkareh) [inline src/terminal-screen.c:1695]: Maybe set it more explicitly to `FLAVOR_AS_IS`, or add it as a comment:
```
  int url_flavor = 0; /* FLAVOR_AS_IS */
```

File: src/terminal-screen.c [hunk 5]
```diff
@@ -1684,33 +1690,57 @@ terminal_screen_button_press (GtkWidget      *widget,
...
 
 	state = event->state & gtk_accelerator_get_default_mod_mask ();
 
-	matched_string = terminal_screen_check_match (screen, (GdkEvent*)event, &matched_flavor);
+	hyperlink = terminal_screen_check_hyperlink (screen, (GdkEvent*)event);
+
+	if (hyperlink != NULL &&
+		(event->button == 1 || event->button == 2) &&
+		(state & GDK_CONTROL_MASK))
+	{
```
Comment (vkareh) [inline src/terminal-screen.c:1705]: I don't like the structure of this function. It conflates the `url` or `hyperlink` values with button press itself in the conditional statements, yet this is a button press callback function. I would suggest refactoring this to something like:
```c
	gboolean handled = FALSE;

	hyperlink = terminal_screen_check_hyperlink (screen, (GdkEvent*)event);
	url = terminal_screen_check_match (screen, (GdkEvent*)event, &url_flavor);

	if (event->button == 1 || event->button == 2) && (state & GDK_CONTROL_MASK))
	{
		if (hyperlink != NULL) // ...
		else if (url != NULL) // ...
	}

	if (event->button == 3 && (state & (GDK_SHIFT_MASK | GDK_CONTROL_MASK | GDK_MOD1_MASK)) == 0)
	{ /* ... popup stuff ... */ }

	g_free (hyperlink);
	g_free (url);

	if (handled)
		return TRUE;

	/* default behavior is to let the terminal widget deal with it */
	if (button_press_event)
		return button_press_event (widget, event);

	return FALSE;
```

It's a bit cleaner in my opinion, and removes all those stray `g_free` calls...

**Comment by vkareh** [comment]
@lukefromdc 

> I don't really understand this, so will leave it for other team members to work on. I DO see a possible issue in your linked explanation though: autocompleting URLs could lead people to sites they don't ever want to connect to (e.g Facebook, Tiktok) or even to malware sites. Anything networked requires serious security, and coding it is out of my experience beyond blocking such sites on my machines.

This just affects rendering. What this does is expose functionality already available in VTE: hypertext linking. We already do the same thing with plain links, this just adds the hypertext portion.

Screenshot at 2023-03-08 12-24-55

Screenshot at 2023-03-08 12-25-14

Screenshot at 2023-03-08 12-25-27

File: src/terminal-screen.c (hunk 3 above)
Comment (geo-stark) [reply src/terminal-screen.c:1710]: as I found later - url_flavor is not assigned for hyperlinks

File: src/terminal-screen.h (hunk 1 above)
Comment (vkareh) [reply src/terminal-screen.h:38]: I mean the first: move renaming `TerminalURLFlavor` to a separate commit.

File: src/terminal-util.c [hunk 6]
```diff
@@ -707,6 +708,56 @@ terminal_util_add_proxy_env (GHashTable *env_table)
...
+		return NULL;
+
+	unesc = g_uri_unescape_string(uri, NULL);
+	if (unesc == NULL)
+		unesc = g_strdup(uri);
+
+	if (g_ascii_strncasecmp(unesc, "ftp://", 6) == 0 ||
+		g_ascii_strncasecmp(unesc, "http://", 7) == 0 ||
+		g_ascii_strncasecmp(unesc, "https://", 8) == 0 ||
+		g_ascii_strncasecmp(unesc, "mailto://", 9) == 0)
```
Comment (vkareh) [inline src/terminal-util.c:735]: mailto doesn't use slashes, should be `"mailto:", 7`, but I don't know if mail addresses are ever IDN. I guess they can be anyway. I can confirm against the RFC next time I have a chance.

File: src/terminal-screen.h (hunk 1 above)
Comment (geo-stark) [reply src/terminal-screen.h:38]: BTW all renamings are in separate commit 285a0fcc8f29a7cc5ccb9d8f7796c9f2f68ff961 already
I've just mentioned TerminalURLFlavor renaming in commit message.

File: src/terminal-util.c (hunk 6 above)
Comment (geo-stark) [reply src/terminal-util.c:735]: Year, my fault. I took a looked at the code more attentive and it won't be work correctly with email addresses anyway. I'll fix it in a separate patch

File: src/terminal-screen.h (hunk 1 above)
Comment (vkareh) [reply src/terminal-screen.h:38]: Ah, right, my bad, I just saw the refactoring commit 👍

**Comment by raveit65** [comment]
@vkareh @geo-stark 
What is the status of this?

**Comment by vkareh** [comment]
This looks good to me. I just did a quick review of the code and tested it and it still works as I remember. I'm fine if we merge this.

**Comment by raveit65** [comment]
@vkareh 
Any chance to approve it :)

**Comment by geo-stark** [comment]
> @vkareh @geo-stark What is the status of this?

I responded to all comments regarding the patches. If community has other questions \ suggestions I'm ready to answer it

**Comment by raveit65** [comment]
> > I have no idea how to test this, would need to be able to do so without connection to unknown/unsafe sites
> 
> Overall information about this feature: https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda some quick tests:
> 
>     1. printf '\e]8;;[https://ubuntu.com/\e\\Ubuntu](https://ubuntu.com/%5Ce%5C%5CUbuntu) 21.10\e]8;;\e\\n'
> 
>     2. ls --hyperlink=always

A better page for informations seems to be here https://github.com/Alhadis/OSC8-Adoption

**Review by raveit65 — APPROVED** [review]
LGTM,
works fine with those example commands.
`printf '\033]8;;http://example.com\033\\This is a link\033]8;;\033\\\n'`
`ls --hyperlink=always`

File: src/terminal-window.c [hunk 7]
```diff
@@ -1365,6 +1365,38 @@ handle_tab_droped_on_desktop (GtkNotebook *source_notebook,
...
+                               TerminalWindow *window)
+{
+    TerminalWindowPrivate *priv = window->priv;
+    TerminalScreenPopupInfo *info = priv->popup_info;
+
+    if (info == NULL)
+        return;
+
+    terminal_util_open_url (GTK_WIDGET (window), info->hyperlink, FLAVOR_AS_IS,
+                          gtk_get_current_event_time ());
```
Comment (raveit65) [inline src/terminal-window.c:1379]: Indents
```suggestion
    terminal_util_open_url (GTK_WIDGET (window), info->hyperlink, FLAVOR_AS_IS,
                            gtk_get_current_event_time ());
```

Comment (geo-stark) [reply src/terminal-window.c]: I fixed specified indents as well as few others
